lib/stdlib/src/lists.erl

```diff
@@ -96,58 +96,52 @@ An example of a typical ordering function is less than or equal to: `=</2`.
 
 %% Shadowed by erl_bif_types: lists:keyfind/3
 -doc """
-Searches the list of tuples `TupleList` for a tuple whose `N`th element compares
-equal to `Key`. Returns `Tuple` if such a tuple is found, otherwise `false`.
+Searches the list `List` for a tuple whose `N`th element compares
+equal to `Key`. Returns `Elem` if such a tuple is found, otherwise `false`.
 """.
--spec keyfind(Key, N, TupleList) -> Tuple | false when
+-spec keyfind(Key, N, List) -> Elem | false when
       Key :: term(),
       N :: pos_integer(),
-      TupleList :: [Tuple],
-      Tuple :: tuple().
+      List :: [Elem].
```

Comment:
In the current state of the type system, we lack bounded quantification. If we were to accept this type as written, our users will lose type information because not all type systems can infer that `Elem` must be a subtype of `tuple()`, more specifically, of the tuple passed as argument in the list.

I understand that is the more precise one can get, but it also means losing type information.
If we add bounded quantification, we could say:

```erlang
-spec keyfind(Key, N, List) -> Elem | false when
    Key :: term(),
    N :: pos_integer(),
    List :: [Elem],
    Elem <: tuple().
```

In this case, some type systems may choose `Elem = tuple()` and move on, and strong type systems may figure out the exact tuple value expected. For now, I do not think we can go with this level of generic programming